core/Pos/Core/Arbitrary.hs

```diff
@@ -318,26 +332,42 @@ instance Arbitrary G.StakeDistribution where
       , return G.ExponentialStakes
       , G.ExplicitStakes <$> arbitrary
       ]
+    shrink = genericShrink
 
 ----------------------------------------------------------------------------
 -- Arbitrary miscellaneous types
 ----------------------------------------------------------------------------
 
 instance Arbitrary Millisecond where
     arbitrary = fromMicroseconds <$> choose (0, 600 * 1000 * 1000)
+    shrink = fmap fromIntegral . shrink . toInteger
```

Comment:
How do you know that this is safe? What if the shrinking strategy for `Integer` changes and we start to get under/overflow? I'm not saying that this will happen, but I'm saying that this code is correct by coincidence.

What you should do instead is use `shrinkIntegral` and newtype wrapping/unwrapping. This criticism applies to both `Millisecond` and `Microsecond` instances (slightly below).